[Distributed] Optimize checkpoint resharding with sweep-line algorithm

mansiag05 opened the pull request:
Replace nested loop in `create_read_items_for_chunk_list` with sweep-line algorithm for finding overlapping chunks during checkpoint resharding.

The optimization:
- Sort chunks by position on sweep dimension (largest range)
- Use `bisect_right` and `insort` to maintain active set
- Only compare chunks that could potentially overlap

This reduces overhead when resharding checkpoints between different parallelization strategies.

Test coverage already exists at:
test/distributed/checkpoint/test_planner.py::TestPlannerHelpers::test_create_read_item_from_chunks

Fixes #169113

mansiag05 committed:
[Distributed] Optimize checkpoint resharding with sweep-line algorithm

Replace nested loop in `create_read_items_for_chunk_list`
with sweep-line algorithm for finding overlapping chunks
during checkpoint resharding.

The optimization:
- Sort chunks by position on sweep dimension (largest range)
- Use `bisect_right` and `insort` to maintain active set
- Only compare chunks that could potentially overlap

This reduces overhead when resharding checkpoints between different
parallelization strategies.

Test coverage already exists at:
test/distributed/checkpoint/test_planner.py::TestPlannerHelpers::test_create_read_item_from_chunks

Skylion007 commented on torch/distributed/checkpoint/planner_helpers.py:
```diff
@@ -28,6 +28,8 @@
     WriteItem,
     WriteItemType,
 )
+from bisect import bisect_right, insort
```
Linter is going to complain since this is from the stdlib and the import isn't properly sorted once the CI is back up.

mansiag05 replied:
Thanks for pointing it out Aaron. Ran the lint locally, and updated the file.

mansiag05 committed:
Correcting the lint issue.

Skylion007 commented on torch/distributed/checkpoint/planner_helpers.py:
```diff
@@ -270,11 +271,63 @@ def create_read_items_for_chunk_list(
+        max_size = 0
+        for dim in range(num_dims):
+            dim_size = max(
+                max(chunk.offsets[dim] + chunk.sizes[dim] for chunk in local_chunks),
```
These two max ops are on identical generators with different args? Just call them on itertools.chain(local_chunks, saved_chunks) of the two items?

mansiag05 replied:
Corrected the nested max calls using itertools.chain

mansiag05 committed:
Flatten nested max() calls using itertools.chain

mansiag05 committed:
[DCP] Fix checkpoint loading for 0-dimensional tensors

The sweep line algorithm was accessing offsets[sweep_dim]
which raises an IndexError for 0-d tensors since their offsets tuple is empty.

Fix by pre-computing bounds for all chunks before the sweep line loop.
For 0-d tensors, use (0, 1) bounds so all chunks overlap in the sweep
dimension, then let _check_shard_metadata_pair_overlap handle the
actual overlap check (which correctly returns True for 0-d tensors).

mansiag05 commented:
Looks like the CI issues are unrelated and should be resolved on re-triggering. Could anyone please retrigger the CI on the PR. 
@cyyever @Skylion007 @wconstab

cyyever commented:
@pytorchbot rebase

mansiag05 commented:
Hello @cyyever, thanks for the approval. 
The failing jobs looks like are not related to the change in the PR, one is openreg timeout failure and other seems like MPS device was not available on the machine on which the job ran. Hence it failed with import issue.

Trying to re-trigger the CI.

wconstab commented:
If you're sure the fails are unrelated you can use the -i flag for merge bot, else just rebase on viable/strict

mansiag05 commented:
@pytorchbot rebase -b viable/strict
